docs/developer_overview.md

```diff
@@ -0,0 +1,527 @@
+```
+
+The dispatch object's `Invoke` method is then called with proper policy:
+
+```cpp
+template <typename InputIteratorT, 
+          /* ... */
+          typename SelectedPolicy = DefaultTuning>
+struct DispatchReduce : SelectedPolicy 
+{
+    
+template <typename ActivePolicyT>
+CUB_RUNTIME_FUNCTION __forceinline__
+cudaError_t Invoke()
+{
+  using MaxPolicyT = typename DispatchSegmentedReduce::MaxPolicy;
+
+  return InvokePasses<ActivePolicyT /* (1) */>(
+    DeviceReduceKernel<MaxPolicyT /* (2) */, InputIteratorT /* ... */>);
+}
+    
+};
+```
+
+This is where the actual work happens. Note how the kernel is used against `MaxPolicyT` (2) while the kernel invocation part uses `ActivePolicyT` (1). This is an important part:
+
+```cpp
+template <typename ChainedPolicyT /* ... */ >
+__launch_bounds__(int(ChainedPolicyT::ActivePolicy::ReducePolicy::BLOCK_THREADS)) 
+__global__ void DeviceReduceKernel(InputIteratorT d_in /* ... */) 
+{
+  // Thread block type for reducing input tiles
+  using AgentReduceT =
+      AgentReduce<typename ChainedPolicyT::ActivePolicy::ReducePolicy,
+                  InputIteratorT, OutputIteratorT, OffsetT, ReductionOpT>;
+
+  // Shared memory storage
+  __shared__ typename AgentReduceT::TempStorage temp_storage;
+
+  // Consume input tiles
+  OutputT block_aggregate =
+      AgentReduceT(temp_storage, d_in, reduction_op).ConsumeTiles(even_share);
+
+  // Output result
+  if (threadIdx.x == 0) {
+    d_out[blockIdx.x] = block_aggregate;
+  }
+}
+```
+
+The kernel gets compiled for each PTX version that was provided to the compiler. During the device pass, `ChainedPolicy` compares `CUDA_ARCH` against the template parameter to select `ActivePolicy` type alias. During the host pass, `Invoke` is compiled for each architecture in the tuning list. If we use `ActivePolicy` instead of `MaxPolicy` as a kernel template parameter, we will compile `O(N^2)` kernels instead of `O(N)`. 
+
+Finally, the tuning looks like:
+
+```cpp
+template <typename InputT /* ... */>
+struct DeviceReducePolicy 
+{
+  /// SM35
+  struct Policy350 : ChainedPolicy<350, Policy350, Policy300> {
+    typedef AgentReducePolicy<256, 20, InputT, 4, BLOCK_REDUCE_WARP_REDUCTIONS,
+                              LOAD_LDG>
+        ReducePolicy;
+  };
+
+  /// SM60
+  struct Policy600 : ChainedPolicy<600, Policy600, Policy350> {
+    typedef AgentReducePolicy<256, 16, InputT, 4, BLOCK_REDUCE_WARP_REDUCTIONS,
+                              LOAD_LDG>
+        ReducePolicy;
+  };
+
+  /// MaxPolicy
+  typedef Policy600 MaxPolicy;
+};
+```
+
+The kernels in the dispatch layer shouldn't contain a lot of code. Usually, the functionality is extracted into the agent layer. All the kernel does is derive the proper policy type, unwrap the policy to initialize the agent and call one of its `Consume`/`Process` methods. Agents are frequently reused by unrelated device-scope algorithms. 
+
+#### Temporary storage usage
+
+It's safe to reuse storage in the stream order:
```

Comment:
That sentence is in contrast to the example, where the last call is illegal. We should make it clear that we need to potentially call `cudaStreamSynchronize(stream)`

Reply:
The last call is not in stream order, so I don't see any contradiction here